cladecombiner/aggregator.py

```diff
@@ -0,0 +1,278 @@
+            The taxa we wish to aggregate.
+        """
+        self.input_taxa = list(in_taxa)
+        "The taxa we wish to aggregate."
+        self.stack = copy(self.input_taxa)
+        "Taxa that still need to be aggregated."
+        self.messy_map = {}
+        "A record of the entire history of aggregation of all taxa."
+
+    def _valid_agg_step(self, mapping: TaxonMapping):
+        """
+        Check that this proposed merge move is acceptable
+        """
+        unknown = [k for k in mapping.map.keys() if k not in self.stack]
+        if len(unknown) > 0:
+            raise RuntimeError(
+                f"Tried to map {unknown} which is not in current stack {self.stack}."
+            )
+
+    def _valid_mapping(self):
+        """
+        Checks that all input taxa have been mapped exactly once.
+        """
+        in_str = [taxon.name for taxon in self.input_taxa]
+        map = self.map()
+        if set(map.keys()) != set(in_str):
+            raise RuntimeError(
+                "Mismatch between mapped taxa and input taxa. Input taxa are: "
+                + str(in_str)
+                + " but mapped taxa are "
+                + str(map.keys())
+            )
+        tab = table(map)
+        if not all(v == 1 for v in tab.values()):
+            raise RuntimeError(
+                "Found following taxa mapped more than once: "
+                + str([k for k, v in tab.items() if v > 1])
+            )
+
+    def aggregate(self):
+        """
+        Perform aggregation. When done, results may be collected with self.map().
+
+        In the interim, self.messy_map is populated with the results of aggregation.
+        """
+        while self.stack:
+            step = self.next_agg_step()
+            self._valid_agg_step(step)
+            self.resolve_agg_step(step)
+
+        self._valid_mapping()
+
+    @abstractmethod
+    def next_agg_step(self) -> TaxonMapping:
+        """
+        Find the next available step in the aggregation process.
+        """
+        pass
+
+    def map(self) -> dict[str, str]:
+        """
+        Create clean {Taxon from : Taxon to} aggregation dictionary.
+        """
+        if len(self.stack) > 0:
+            self.aggregate()
+        taxon_map = {}
+        for taxon_from in self.input_taxa:
+            taxon_to = taxon_from
+            while taxon_to in self.messy_map.keys():
+                if taxon_to == self.messy_map[taxon_to]:
+                    # Self-mappings are possible
+                    break
+                taxon_to = self.messy_map[taxon_to]
+            taxon_map[taxon_from.name] = taxon_to.name
+        return taxon_map
+
+    def resolve_agg_step(self, mapping: TaxonMapping):
+        for k, v in mapping.map.items():
+            self.messy_map[k] = v
+            self.stack.remove(k)
+
+        for k, v in mapping.done.items():
+            if not v:
+                self.stack.append(k)
+
+
+class FixedAggregator(Aggregator):
```

Comment:
To check my understanding, is this aggregation just one step? You assign everything all in one go?

But it looks like we think about aggregation steps as function calls in different methods. Like, it's not the case that a "basic" aggregation is, in any way, a series of "fixed" aggregations?

Maybe I need to keep reading

Reply:
No, this is a fair point.

Both of these are _terrible_ fits for the step-by-step process I built. Our notes said this was Right and Proper, but reality keeps coming short of that.